File: mac/Sources/YTRec/Models/YtURL.swift
import Foundation

/// YouTube 網址解析（純邏輯，可測試）
enum YtURL {
    static func isProbablyYouTube(_ raw: String) -> Bool {
        guard let url = URL(string: raw.trimmingCharacters(in: .whitespacesAndNewlines)),
              let host = url.host?.lowercased() else { return false }
        return host.contains("youtube.com") || host.contains("youtu.be")
    }

    /// 從各種 YouTube 網址型態抽出 11 碼影片 ID
    static func videoID(_ raw: String) -> String? {
        let s = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let url = URL(string: s) else { return nil }
        let idPattern = "^[A-Za-z0-9_-]{11}$"
        func valid(_ c: String?) -> String? {
            guard let c, c.range(of: idPattern, options: .regularExpression) != nil else { return nil }
            return c
        }
        if let host = url.host?.lowercased(), host.contains("youtu.be") {
            return valid(url.pathComponents.dropFirst().first)
        }
        if let comps = URLComponents(url: url, resolvingAgainstBaseURL: false),
           let v = comps.queryItems?.first(where: { $0.name == "v" })?.value,
           let id = valid(v) {
            return id
        }
        let parts = url.pathComponents
        for (i, p) in parts.enumerated() where ["live", "shorts", "embed", "v"].contains(p) {
            if i + 1 < parts.count, let id = valid(parts[i + 1]) { return id }
        }
        return nil
    }
}

File: windows/YtRec.Core/YtUrl.cs
using System.Text.RegularExpressions;

namespace YtRec.Core;

/// <summary>YouTube URL parsing (pure logic). Mirrors mac YtURL.swift.</summary>
public static class YtUrl
{
    private const string IdPattern = "^[A-Za-z0-9_-]{11}$";

    public static bool IsProbablyYouTube(string raw)
    {
        if (!Uri.TryCreate(raw.Trim(), UriKind.Absolute, out var url)) return false;
        var host = url.Host.ToLowerInvariant();
        return host.Contains("youtube.com") || host.Contains("youtu.be");
    }

    /// <summary>Extract the 11-char video ID from any YouTube URL form; null if none.</summary>
    public static string? VideoId(string raw)
    {
        if (!Uri.TryCreate(raw.Trim(), UriKind.Absolute, out var url)) return null;

        static string? Valid(string? c) =>
            c != null && Regex.IsMatch(c, IdPattern) ? c : null;

        var host = url.Host.ToLowerInvariant();
        if (host.Contains("youtu.be"))
            return Valid(url.AbsolutePath.Split('/', StringSplitOptions.RemoveEmptyEntries).FirstOrDefault());

        // ?v= takes precedence over path segments
        foreach (var kv in url.Query.TrimStart('?').Split('&', StringSplitOptions.RemoveEmptyEntries))
        {
            var eq = kv.IndexOf('=');
            if (eq > 0 && kv[..eq] == "v")
            {
                var id = Valid(Uri.UnescapeDataString(kv[(eq + 1)..]));
                if (id != null) return id;
            }
        }

        var parts = url.AbsolutePath.Split('/', StringSplitOptions.RemoveEmptyEntries);
        for (var i = 0; i < parts.Length; i++)
        {
            if (parts[i] is "live" or "shorts" or "embed" or "v" && i + 1 < parts.Length)
            {
                var id = Valid(parts[i + 1]);
                if (id != null) return id;
            }
        }
        return null;
    }
}
